relays/substrate/src/millau_headers_to_rialto.rs

```diff
@@ -96,8 +101,71 @@ impl TargetClient<MillauHeadersToRialto> for RialtoTargetClient {
 	}
 }
 
+/// Maintainment procedure of Millau -> Rialto headers.
+struct MillauHeadersToRialtoMaintain {
+	millau_justifications: Arc<Mutex<JustificationsSubscription>>,
+}
+
+#[async_trait]
+impl SyncMaintain<MillauHeadersToRialto> for MillauHeadersToRialtoMaintain {
+	async fn maintain(&self, sync: &mut HeadersSync<MillauHeadersToRialto>) {
+		poll_fn(|context| {
+			let mut millau_justifications = match self.millau_justifications.try_lock() {
+				Some(millau_justifications) => millau_justifications,
+				None => {
+					// this should never happen, as we use single-thread executor
+					log::warn!(target: "bridge", "Failed to lock Millau justifications mutex");
+					return Poll::Ready(());
+				}
+			};
+			loop {
+				let maybe_next_millau_justification = millau_justifications.next();
+				futures::pin_mut!(maybe_next_millau_justification);
+
+				let maybe_next_millau_justification = maybe_next_millau_justification.poll_unpin(context);
+				match maybe_next_millau_justification {
+					Poll::Ready(millau_justification) => {
+						// TODO: we only need last justification before incomplete header - no need to submit all
+						// pending justifications
+
+						// TODO: move justification decode to primitives && decode at Client fn
+						use pallet_substrate_bridge::decode_justification_target;
+						let justification_target =
+							decode_justification_target::<HeaderOf<Millau>>(&millau_justification);
+						match justification_target {
+							Ok((target_hash, target_number)) => {
+								let target = HeaderId(target_number, target_hash);
+								if sync.headers().requires_completion_data(&target) {
+									sync.headers_mut()
+										.completion_response(&target, Some(millau_justification));
+									continue;
+								}
+								if sync.headers().status(&target) == HeaderStatus::Synced {
+									// TODO: https://github.com/paritytech/parity-bridges-common/issues/209
+									// submit justification to Rialto
+								}
+							}
+							Err(error) => {
+								log::warn!(
+									target: "bridge",
+									"Failed to decode justification from Millau: {:?}",
+									error,
+								);
+							}
+						}
+					}
+					Poll::Pending => break,
```

Comment:
Is this correct? Current `maybe_next_millau_justification` would be lost?

Reply:
I can't say for sure now - will check before making it ready for review :)

Reply:
Now I'm pretty sure nothing is lost :)